### PR TITLE
MAHOUT-1988 Make Native Solvers Scala 2.11 Complient

### PR DESCRIPTION
**Pull request by rawkintrevo**
### Purpose of PR:
Please give a short description of what this PR is for.

### Important ToDos
Please mark each with an "x"
- [x] A JIRA ticket exists (if not, please create this first)[https://issues.apache.org/jira/browse/ZEPPELIN/]
- [x] Title of PR is "MAHOUT-XXXX Brief Description of Changes" where XXXX is the JIRA number.
- [x] Assigned JIRA to self
- [x] Successfully built and ran all unit tests, verified that all tests pass locally.

If all of these things aren't complete, but you still feel it is
appropriate to open a PR, please add [WIP] after MAHOUT-XXXX before the
descriptions- e.g. "MAHOUT-XXXX [WIP] Description of Change"

Does this change break earlier versions?
no

Is this the beginning of a larger project for which a feature branch should be made?
no

Issue was- as error message stated, default arguments not allowed in multiple constructors- scala gets confuzzled.

### CODE REVIEW

**Comment by rawkintrevo** [comment]
fwiw- this doesn't actually fix the dependency issue. I think I'll make a new jira ticket for this, then down grade mahout-1988 to trivial

**Comment by andrewpalumbo** [comment]
getting: ```*** RUN ABORTED ***
  java.lang.NoSuchMethodError: scala.runtime.IntRef.create(I)Lscala/runtime/IntRef;
  at org.apache.mahout.viennacl.opencl.package$.repackCSR(package.scala:227)
  at org.apache.mahout.viennacl.opencl.package$.toVclCmpMatrixAlt(package.scala:120)
  at org.apache.mahout.viennacl.opencl.GPUMMul$.org$apache$mahout$viennacl$opencl$GPUMMul$$gpuRWCW(GPUMMul.scala:161)
  at org.apache.mahout.viennacl.opencl.GPUMMul$.org$apache$mahout$viennacl$opencl$GPUMMul$$jvmRWRW(GPUMMul.scala:208)
  at org.apache.mahout.viennacl.opencl.GPUMMul$$anonfun$12.apply(GPUMMul.scala:78)
  at org.apache.mahout.viennacl.opencl.GPUMMul$$anonfun$12.apply(GPUMMul.scala:78)
  at org.apache.mahout.viennacl.opencl.GPUMMul$.apply(GPUMMul.scala:127)
  at org.apache.mahout.viennacl.opencl.GPUMMul$.apply(GPUMMul.scala:33)
  at org.apache.mahout.math.scalabindings.RLikeMatrixOps.$percent$times$percent(RLikeMatrixOps.scala:37)
  at org.apache.mahout.opencl.viennacl.ViennaCLSuiteVCL$$anonfun$2.apply$mcV$sp(ViennaCLSuiteVCL.scala:92)``` when running `mvn test` in the viennacl package.

File: viennacl/src/main/scala/org/apache/mahout/viennacl/opencl/javacpp/DenseColumnMatrix.scala [hunk 1]
```diff
@@ -36,12 +36,24 @@ import org.bytedeco.javacpp.annotation._
...
+    this(false)
+    allocate(nrow, ncol, new Context())
+  }
+
+  def this(nrow: Int, ncol: Int, ctx: Context) {
     this(false)
     allocate(nrow, ncol, ctx)
   }
 
-  def this(data: DoublePointer, nrow: Int, ncol: Int, ctx: Context = new Context(Context.MAIN_MEMORY)) {
```
Comment (andrewpalumbo) [inline viennacl/src/main/scala/org/apache/mahout/viennacl/opencl/javacpp/DenseColumnMatrix.scala]: getting: ```*** RUN ABORTED *** java.lang.NoSuchMethodError: scala.runtime.IntRef.create(I)Lscala/runtime/IntRef; at org.apache.mahout.viennacl.opencl.package$.repackCSR(package.scala:227) at org.apache.mahout.viennacl.opencl.package$.toVclCmpMatrixAlt(package.scala:120) at org.apache.mahout.viennacl.opencl.GPUMMul$.org$apache$mahout$viennacl$opencl$GPUMMul$$gpuRWCW(GPUMMul.scala:161) at org.apache.mahout.viennacl.opencl.GPUMMul$.org$apache$mahout$viennacl$opencl$GPUMMul$$jvmRWRW(GPUMMul.scala:208) at org.apache.mahout.viennacl.opencl.GPUMMul$$anonfun$12.apply(GPUMMul.scala:78) at org.apache.mahout.viennacl.opencl.GPUMMul$$anonfun$12.apply(GPUMMul.scala:78) at org.apache.mahout.viennacl.opencl.GPUMMul$.apply(GPUMMul.scala:127) at org.apache.mahout.viennacl.opencl.GPUMMul$.apply(GPUMMul.scala:33) at org.apache.mahout.math.scalabindings.RLikeMatrixOps.$percent$times$percent(RLikeMatrixOps.scala:37) at org.apache.mahout.opencl.viennacl.ViennaCLSuiteVCL$$anonfun$2.apply$mcV$sp(ViennaCLSuiteVCL.scala:92)```

when running `mvn test` in the viennacl package.

Comment (andrewpalumbo) [inline viennacl/src/main/scala/org/apache/mahout/viennacl/opencl/javacpp/DenseColumnMatrix.scala]: hmm. when running `mvn test` on a scala 2.11.2 build it seems to bee looking for ViennCL built for 2.10:  `mahout-native-viennacl_2.10 `
```
[INFO] Copying 3 resources
[INFO] 
[INFO] --- scala-maven-plugin:3.2.0:testCompile (scala-test-compile) @ mahout-native-viennacl_2.10 ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- maven-compiler-plugin:3.3:testCompile (default-testCompile) @ mahout-native-viennacl_2.10 ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- maven-surefire-plugin:2.18.1:test (default-test) @ mahout-native-viennacl_2.10 ---
[INFO] Tests are skipped.
[INFO] 
[INFO] --- scalatest-maven-plugin:1.0:test (test) @ mahout-native-viennacl_2.10 ---
```
Maybe the scalatest plugin in scala version is hardcoded in a pom somewere?.

Comment (andrewpalumbo) [inline viennacl/src/main/scala/org/apache/mahout/viennacl/opencl/javacpp/DenseColumnMatrix.scala]: @rawkintrevo I have no problems building this for scala-2.11, with E.g.: `mvn clean install -Dscala.version=2.11.2 -Dscala.compat.version=2.11 -Pviennacl -Phadoop2  -DskipTests`` *note* the scala version is @ 2.11 which I think wee may need to make our cap due to the Breeze vector cap @scala 2.11.2 :
```
 [WARNING]  org.scalanlp:breeze_2.11:0.11.2 requires scala version: 2.11.4 
```
```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] Mahout Build Tools ................................. SUCCESS [  3.189 s]
[INFO] Apache Mahout ...................................... SUCCESS [  0.080 s]
[INFO] Mahout Math ........................................ SUCCESS [ 13.226 s]
[INFO] Mahout HDFS ........................................ SUCCESS [  3.310 s]
[INFO] Mahout Map-Reduce .................................. SUCCESS [ 21.442 s]
[INFO] Mahout Integration ................................. SUCCESS [  6.183 s]
[INFO] Mahout Examples .................................... SUCCESS [ 21.106 s]
[INFO] Mahout Math Scala bindings ......................... SUCCESS [ 54.202 s]
[INFO] Mahout Spark bindings .............................. SUCCESS [ 51.939 s]
[INFO] Mahout Native VienniaCL OpenCL Bindings ............ SUCCESS [ 48.965 s]
[INFO] Mahout Native VienniaCL OpenMP Bindings ............ SUCCESS [ 25.782 s]
[INFO] Mahout Release Package ............................. SUCCESS [  2.567 s]
[INFO] Mahout H2O backend ................................. SUCCESS [ 22.381 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 04:34 min
[INFO] Finished at: 2017-06-23T16:39:21-07:00
[INFO] Final Memory: 132M/1124M
[INFO] -----------------------------------
```

tests are failing for me though in `viennacl`.  I'll to dig deeper.

Comment (rawkintrevo) [inline viennacl/src/main/scala/org/apache/mahout/viennacl/opencl/javacpp/DenseColumnMatrix.scala]: @andrewpalumbo thanks for the review

If you look at line 49 and 56 below- this is what replaces.  Line 49 lets you call the constructor with out a default context (same as what happened here) line 56 lets you explicitly pass a context. 

Re: Scala 2.11.2, I made a comment on MAHOUT-1975.  We probably need to figure out where the dependency comes from and try to work around it. 

Will also dig in on my end- viennacl always fails for me bc I don't have strong enough GPUs- but results should be same for ViennaCL-OMP.

Comment (rawkintrevo) [inline viennacl/src/main/scala/org/apache/mahout/viennacl/opencl/javacpp/DenseColumnMatrix.scala]: @andrewpalumbo can you try

`mvn test -Dscala.version=2.11.8 -Dscala.compat.version=2.11`

The reason I believe this works is- 
```
<dependency>
      <groupId>org.scalatest</groupId>
      <artifactId>scalatest_${scala.compat.version}</artifactId>
    </dependency>
```
where scala-compat-version is by default 2.10.  So when you `mvn test` you are doing so with scala-test_2.10, and that's where its losing its mind.

(Also explains why these were passing in Travis)

Comment (andrewpalumbo) [inline viennacl/src/main/scala/org/apache/mahout/viennacl/opencl/javacpp/DenseColumnMatrix.scala]: rebuilding abd testing now with the correct scala flags.. noticed this in spark/dependency-reduced-pom.xmlpom.xml:  [INFO] Reading assembly descriptor: src/main/assembly/dependency-reduced.xml

```
[WARNING] The following patterns were never triggered in this artifact inclusion filter:
o  'org.apache.mahout:mahout-native-viennacl_2.10'
o  'org.apache.mahout:mahout-native-viennacl-omp_2.10'

[INFO] Building jar: /home/andy/sandbox/mahout_cuda/spark/target/mahout-spark_2.11-0.13.1-SNAPSHOT-dependency-reduced.jar
```
*easyhack*: add
to `spark/dependency reduced.xml`:
```
  'org.apache.mahout:mahout-native-viennacl_2.11'
  'org.apache.mahout:mahout-native-viennacl-omp_2.11'
```
though that pom should honor the `{scala.compat.version}`

Comment (andrewpalumbo) [inline viennacl/src/main/scala/org/apache/mahout/viennacl/opencl/javacpp/DenseColumnMatrix.scala]: @rawkintrevo: 
```
$MAHOUT_HOME/viennacl$ mvn  test -Dscala.version=2.11.2 -Dscala.compat.version=2.11 -Pviennacl  
```
(I'd tried built with 2.11.2)
tests pass: 
```
All tests passed.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 01:36 min
[INFO] Finished at: 2017-06-24T13:49:59-07:00
[INFO] Final Memory: 20M/296M
[INFO] ------------------------------------------------------------------------
```
Sweet- I'd like to try this in Pseudo cluster mode as well, this afternoon.

Comment (andrewpalumbo) [inline viennacl/src/main/scala/org/apache/mahout/viennacl/opencl/javacpp/DenseColumnMatrix.scala]: If you'd like to merge this first i'd be +1 on that.  Then testing on pseudo cluster could be a separate issue.   The `spark/dependency reduced.xml` needs to be updated.

Comment (andrewpalumbo) [inline viennacl/src/main/scala/org/apache/mahout/viennacl/opencl/javacpp/DenseColumnMatrix.scala]: @rawkintrevo lgtm - ship it! we swill still need to do some testing against other spark versions though.  As well, there are some other issues that I've noticed, but can be addressed in other PRs on top of this anyways, eg. #327.

Comment (rawkintrevo) [inline viennacl/src/main/scala/org/apache/mahout/viennacl/opencl/javacpp/DenseColumnMatrix.scala]: fwiw- travis now runs all unit tests in pseudocluster. just fyi- of spark 1.6, 2.0, 2.1

Comment (andrewpalumbo) [inline viennacl/src/main/scala/org/apache/mahout/viennacl/opencl/javacpp/DenseColumnMatrix.scala]: +1 - I'd forgotten that last night.

**Pull request closed without merging**